### ISSUE TITLE
auth.signUp() doesn't error for existing accounts - security vulnerability

### ISSUE THREAD

**Issue by CalebLovell** [supabase/auth#1517]
# Bug report

## Describe the bug

`supabase.auth.signUp()` is not erroring for existing accounts. Right now, you can submit an existing email with _any_ incorrect password, and supabase will return you the account metadata (without a jwt).

## To Reproduce

Go to this [example app](https://auth-vanilla-js.vercel.app/)
Sign-up with an email and a password
Log out
Try to sign up again with the same email using any password you want. Try `asdfasdfasdf` if you want!
You will get an alert saying you logged in, but you won't get a working access token. Just the email you submitted.
You can also view the request in the Network tab of the Dev Tools and see metadata about the account, like when it was created and what provider it uses.

## Expected behavior

Attempting to sign up with an existing email should throw an error.

## System information

- Version of supabase-js: [1.2.1]
- Version of Node.js: [14.17.4]

**Comment by kiptoomm** [comment]
As mentioned here: https://github.com/supabase/auth-js/issues/513#issuecomment-1826233042, it seems to me that the latest [docs](https://supabase.com/docs/reference/javascript/auth-signup) have clarified the approach: 

> If signUp() is called for an existing confirmed user:
When both Confirm email and Confirm phone (even when phone provider is disabled) are enabled in [your project](https://supabase.com/dashboard/project/_/auth/providers), an obfuscated/fake user object is returned.
When either Confirm email or Confirm phone (even when phone provider is disabled) is disabled, the error message, User already registered is returned.

It's still not an optimal solution, but at least we can get an explicit error message by disabling the email confirmation. The error response looks like: 

`{"code":422,"error_code":"user_already_exists","msg":"User already registered"}`

**Comment by RowinVanAmsterdam** [comment]
I do not get the magic link in the mailbox when I try to sign up with an already existing user? I only get the fake user data object in return. 

Email service is working, signup emails and sending magic link manually works without a problem.

**Comment by appurist** [comment]
> I propose that we do several things:
> 
> * Document the response behavior clearly [in the docs](https://github.com/supabase/supabase/tree/master/web)
> * Update our [auth examples](https://github.com/supabase/supabase/tree/master/examples) to exhibit the up to date auth flow
> * Update gotrue to [send a recovery email on repeat sign up attempt ](https://github.com/supabase/gotrue/issues/286)
> 
> As always any help on these ones would be hugely appreciated. And thanks again for the active discussion and helping improve the product 🙏.

2.5 years later it looks like:
- this second point has not been done yet. The [example signup](https://github.com/supabase/supabase/blob/5772fbf9119c56ade17fea2a5561ec4e27157aeb/examples/auth/nextjs/app/login/page.tsx#L51) just tells duplicate signup user to **look for the confirmation email** _that is not coming_.
- the third point: I don't know how many duplicate signups you need to request but I have been testing this quite a bit and have not received any emails for a "successful" but duplicate signUp.

I have yet to find any guidance from Supabase on how to handle the user experience and UI flow for when a user attempts a duplicate signup.  The example code just tells them to check for the confirmation email (which is an incorrect UI that misleads the user).

So the official answer here is to create an edge function that uses the admin API to find out if a user exists and then expose that to the anon client for improved UX handling? If we're going to expose that anyway, what's the point of hiding it in an edge function and using the admin API?

I've been huge on FB for a long time now but the handling of this mess is almost enough to make me switch to Firebase where they actually return an error for an error.

**Comment by appurist** [comment]
@hf I understand the strong desire to avoid providing any means for hackers to identify valid email addresses but the resulting UX is horrible and this is a policy that should be determined by each organizations using Supabase, not Supabase itself (globally). The result is a nasty and problematic onboarding experience. I was in your shoes arguing for tighter control until I became responsible for user creation and management at coindesk.com where I had to be more pragmatic because of this was the #1 reason for new users to abandon the onboarding process (and resulting Support nightmare).

What I don't understand is why this cannot simply be an option on the email or general Auth settings, that organizations can choose the behavior of.

(I am now actively working on moving my projects to Firebase due to the onboarding user experience forced on us by Supabase.)

**Comment by hf** [comment]
> What I don't understand is why this cannot simply be an option on the email or general Auth settings, that organizations can choose the behavior of.

It can be an option, but it needs to be built securely and for us to be able to maintain it properly. I understand your predicament and past experience, but in this case it's not just one web property that we're securing but tons. Some want this, some want the other thing.

If we commit to adding an option now, we have to commit to maintain the option until end-of-time. This is not something we're prepared to do at this stage, but is definitely at the forefront of the team's mind. Until we identify a way to do this _properly_ we are not going to jump into quick fixes.

**Comment by SEI-John** [comment]
Being a leader in the industry requires difficult decisions to be made. A lot of this seems to revolve around hesitation and trying to appeal to everyone, and it's not like there hasn't been time to decide on a roadmap for this: it's been nearly 3 years. The result is a lackluster solution for everyone.

**Comment by appurist** [comment]
@hf I'm not expecting a quick fix, as @SEI-John pointed out, it has been years already. I would like Supabase to recognize that this is a business decision to be made by the developer's company, and to commit to ANY solution that seem YOUR team deems reasonable that includes at least some level of pragmatism for this case.  The bottom line is a forced horrible UX for your customers' customers. You need to do _something_ about that.

I've been a professional full-time developer since January 1984 and it seems the focus of the company is not on the ball on this one (at all).  The staunch defense of the status quo has shaken my faith in Supabase, even though I have been a huge fan for years and want to keep using it.

**Comment by appurist** [comment]
There are simpler solutions than making it an option, such as for these "successful" registrations, just send the confirmation email, or one that says "You already have an account; here is a link to login." The problem is the expected email never comes and no messages can indicate to the user that it is not coming, because the Supabase client is **_not_** informed that it was **_not_** a successful registration. 

Just send _any_ reasonable email on a successful registration. We don't really need the option, just that if you're dead set against changing the SB behavior then give us the ability to do so.

You're already sending the confirmation email on a successful new registration; the problem here is you consider a duplicate registration to be successful as well but don't send the email. Solutions that don't affect other Supabase clients:
- For a registration attempt by a known user, send an email that says "You already have an account; here is a link to login." instead of the email confirmation message. or
- Add a field to the result coming black to the client such as "exists: true" so that we can see this. It's less secure for emails but at least we can retain the users being onboarded.

**Comment by appurist** [comment]
For the record, the title is wrong here, **_there is no security issue_** if Supabase just _emailed_ the user to **_login_** if they attempt to register a second time.

A more complicated email with a confirmation key is already being sent for new users; this would just be a canned text email to go to the login page if the user _isn't_ new.

Supabase has such a great product; it's a shame it _forces_ a horrible onboarding experience to some of _our_ users, when there are several solutions readily available, including the one above: backwards-compatible and something all customers would welcome. And for the record, because of this, I've lost my faith in @awalias and Supabase, canceled my subscription and will use an alternative service. I'm _really_ disappointed.

**Comment by hf** [comment]
> For the record, the title is wrong here, there is no security issue if Supabase just emailed the user to login if they attempt to register a second time.

I very much agree with this and I'll personally see what can be done on this point.

**Comment by hf** [comment]
> just that if you're dead set against changing the SB behavior then give us the ability to do so.

This is a strong misrepresentation of the team's intent (and my words). We are not against changing the behavior, and we acknowledge that there is an issue. However, finding a solution to the problem(s) here is honestly more complex than this discussion makes it appear.

Internally this comes up excruciatingly often. We talk about it at least once every two months, sometimes more often than that. So it's not ignored, nor forgotten. We are yet to find a solution that the team likes and can commit to. Some comprehensive solutions (such as offering username+password instead of email+password, where the privacy leak problem comes from) require quite significant changes in the codebase. We are strategically moving in the direction to make emails more optional than they currently are. Once this effort has stabilized we can offer an opinionated approach (use usernames for privacy, use emails for UX).

On the flip side, Supabase Auth is pentested regularly in multiple ways and side-channel leaks for account existence (which is only problematic for email-based authentication) is something that's tested for often.

**Comment by appurist** [comment]
@hf Using a username rather than an email address is not a solution for _this_ problem. It may be a common topic there, but in my experience, that is a hole that you should avoid digging or may never be able to climb out of. That's a different problem. 

The only real problem here, as far as I can tell, is that the repeat signup user _isn't sent an email_.

Is there a problem twith just sending them that email that I'm overlooking? It just needs to inform them of their account status (they already had one) and include a link to the login page.

It avoids all security issues, doesn't require anything of Supabase clients, significantly improves UX for onboarding, and resolves this onboarding case for all users.

**Comment by hf** [comment]
> Is there some problem with just sending them that email that I'm overlooking? It just needs to inform them of their account status (they already had one) and include a link to the login page.

Yes, there is. You're concerned only about your project and your user's experience, but we are concerned about all of Supabase customers and users' experience. 

I can't just add a new email template on a whim. One of our product tenets is that Supabase Auth should be invisible. Once you set it up, you don't need to think about it ever again. Adding a new email template without making it conditional on some configuration option breaks that tenet, and also the promise to our developers that once you set it up one way it will continue to work that way until the end of time. The reason why this is a "breaking change" is that email templates are customizable. Say we've got a customer that has a Spanish-speaking user base and the email templates are all in Spanish. I can't just force a default English template on them. 

So yes now you can say, OK but let's add a configuration option for this. And that's probably the only viable alternative at this point, but we also don't want developing with Supabase Auth to turn into a hundred atomic decisions that we then can't remove or restructure.

Secondly, there's the problem with cost. We can't just start sending more emails without asking for approval from customers. It's them who are paying for email sending so just randomly sending out more emails on every duplicate sign up is irresponsible for us to impose on everyone. We need to consult customers, but also don't want to bother them about Auth at all. 

Third and final, let's say we go with this "fix" of adding a new email template or something similar, there's a config option you turn on, and so on. Who cares about the end user? How do we prevent them from being constantly spammed by attackers, bots, script kiddies, etc? Sending an email on each `signUp` can have serious consequences -- worrying the user too much that someone's trying to get in their account, causing too many similar emails to be received by say Gmail can decrease the reputation of the sending domain making your emails less reliable long-term.

I understand your frustration, and I would be frustrated too. We are trying to do good here for everyone, and this is just going to take more time. I hope I've illustrated how a seemingly simple and "no-brainer" feature request turns into a product decision nightmare at scale.

There exists a workaround with the admin API which you can use to go around this problem should it matter to you for the time being. Please let us know if there's any additional admin API we can provide to make this workaround easier.

I will personally try to see if there's a sufficiently easy way to do this email sending approach, which feels right but requires more serious work, research and strategy to pull of. I can't comment when this will land, in what way, shape or form. When we have updates, we'll share those here.

**Comment by appurist** [comment]
@hf Yes I am primarily focused on the needs of my users (as I should be) but I'm not oblivious to others. I fact, I think there's a fairly fundamental problem here that affects _all_ Supabase clients (who enable email-based registrations).

The first and especially last paragraphs could have been your _entire_ message and perhaps it would have restored some of the confidence lost by 2.5 years of inactivity on a problem affecting the onboarding flow of almost all customers. But the rest of it only highlights an institutional inability to actually _resolve_ such problems. The last paragraph is the proper PR response, but the rest tells me not to expect it. It shows a rigidness there that prevents zooming out to see the Big Picture.

It's okay. I'm not trying to argue for anything in particular here, except I guess _some_ proper resolution (not a server function that just pushes the same problem to another level of indirection) that doesn't need another 2.5 years to achieve. I can accept that this is not likely now, based on your detailed explanation, and act accordingly for my users' needs. (Maybe not Firebase; appwrite.io is looking pretty good so far.)

**Comment by Picoquio** [comment]
Just passing by to express my frustration. This is 100% overstepping. As many have said, this should be a decision made by each organizations using Supabase, not forced by Supabase itself. It's so blatantly obvious that it seems quite a rare hill to die on. You should not force a bad UX on your clients's clients based on a very dogmatic view on privacy. 

> "_This is what I consider the best practice to handle privacy, therefore I'll force it upon everybody else_". 
-Someone in the Supabase team.

Nope. While this approach can be good at maximizing privacy, it comes with a **negative cost of UX** (as we all know and many others have pointed). Your solution isn't a one size fits all solution; hence the organization should decide which side of the coin to pick in this tradeoff.

**Comment by Mr-Pepe** [comment]
@hf Thank you for your response. That was actually very insightful and explains well why any technical changes are not easy to pull off for Supabase as a company with its current values and promises. 

However, I think there is a lot that can be done besides technical changes. 

1. Communication: Had I read this issue a week ago, I would have been very disappointed. It's one of those threads that devolves into a collection of rants, but no reaction from the devs for years. This issue doesn't have the most thumbs up in this repo but the most comments and participants. More regular feedback could help people feel heard and not ignored.
2. API Documentation: The docs currently state `Be aware that if a user account exists in the system you may get back an
   error message that attempts to hide this information from the user.` I had no idea what this was supposed to tell me. What does this statement mean for me as a developer? What does it mean for my users? What is the recommended workflow in that situation?
3. Tutorials: I started working with Supabase Auth through the [Build a User Management App with Flutter](https://supabase.com/docs/guides/getting-started/tutorials/with-flutter) guide and everything seemed so easy. However, going from magic links to email+password felt very confusing. Auto-confirmation is active when developing locally but it is deactivated by default on hosted Supabase, so locally I get an error message signing up twice, but in production that doesn't happen. And then no email gets sent... 

Tackling these issues does not require years to pull off and would probably save a lot of people a lot of time and energy. Better documentation would have saved me a few hours for sure.

**Comment by grokpot** [comment]
Seems like if we can just check `if(data?.user?.identities?.length === 0){` then bad hackers could just read this thread, detect supabase, and check that too.
So, why not just return an error code instead?

**Comment by masda70** [comment]
> The only real problem here, as far as I can tell, is that the repeat signup user isn't sent an email.

@appurist I agree that is definitely a problem in terms of UX. It is typical to expect users to forget that they already have an account, and in such case, the flow is broken.

> There exists a workaround with the admin API which you can use to go around this problem should it matter to you for the time being. Please let us know if there's any additional admin API we can provide to make this workaround easier.

@hf Could you please describe what workaround you had in mind to ensure a sign up for an existing users results in an email?

One easy solution I came up with is to perform a `auth.resetPasswordForEmail` followed by a `auth.signUp` directly from the client using the same `redirectTo` URL for both calls. As far as I understand, this should work reasonably well in terms of user expectations.

**Comment by pariah140** [comment]
This is a very odd design. AWS cognito returns the error message. It seems really dogmatic to have gone down this route and now you are just defending the point because its been done. Is there any way to tell the user the account already exists?

**Comment by appurist** [comment]
@grokpot You cannot rely on that length, as it has different behavior for email vs other providers.

There is no workaround that improves anything over just returning an error here. And forcing us to implement it ourselves in an admin API worker violates one of the key benefits of Supabase: not even needing any project-specific backend code in many cases.

This stance is very disappointing and the refusal to leave the business logic decisions with the businesses has now poisoned my trust of Supabase as a whole. (Don't worry, this will be my last comment on it.)

**Comment by Mykyta-Chernenko** [comment]
UX is not something you can sacrifice
Security is not something you can sacrifice
But they are often go against each other and then the one has to make a choice, like the lifetime of an access token. So it's up to a specific case, company requirements, etc to identify what is the most optimal solution.

This behavior without doubt must be configurable. I'm very surprised it's not now.
We are a local B2B company and we are much more concerned about users forgetting their email accounts then with leaking that an email exist in our system.
Please, make it a config. The developers should be able to decide what is best for them, not what you think is the best for them.

**Comment by oliverjhn** [comment]
I don't understand why this is such a problem. Almost every service I use shows an error message if an user tries to sign up with an already existing email (even supabase themselves).
![image](https://github.com/user-attachments/assets/592a80d4-3e75-4004-9be0-7c28b7f9d714)

**Comment by AbhishekYadav2022** [comment]
Hi

**Comment by oliverjhn** [comment]
I'm not sure if this is the best approach, but this works well enough for me. I have an SQL function that checks if the user's email exists, which I call with RPC from my server action before the user is signed up.

```SQL
DECLARE
  email_exists BOOLEAN;
BEGIN
  SELECT EXISTS (
    SELECT 1
    FROM auth.users
    WHERE email = email_to_check
  ) INTO email_exists;
  
  RETURN email_exists;
END;
```

**Comment by 0xShay** [comment]
> Seems like if we can just check `if(data?.user?.identities?.length === 0){` then bad hackers could just read this thread, detect supabase, and check that too. So, why not just return an error code instead?

Yep I'm using this workaround too, seems to work as intended, thanks - but yeah I don't see the point of sending obfuscated user data when it can easily be debunked in this way. As devs we should get the choice on how to handle this, not have user authentication flow behaviour forced upon us.

**Comment by appurist** [comment]
Once again, the length "workaround" only works for some types of logins. It is not a general workaround.

The only reliable workaround is to create a server-side function that uses the admin API which returns the correct status, then relay that to the client. And if you're going to do that anyway, it accomplishes very little and costs every developer who doesn't want to confuse the user the need to not just follow the documented procedure, but rather to learn of this problem, and create a server-side admin function to expose this status to the client-side caller **_anyway_**.

So where is the savings there? Where is the security that this "feature" brings? Why is Supabase making business decisions for every customer, and forcing a confusing "onboarding" experience for returning users _by default_? Why is it dumb by default? 

If Supabase wants to force this on their developers, at least just continue to send the email on any "successful" registration, not and don't skip it for the "fake successful registrations".  It's confusing to the user, they think they just registered, but it resolves the inability to continue because they don't know they need to login.

**Comment by Picoquio** [comment]
TLDR: you guys at Supabase have created an amazing product, but you're on the wrong here. Do the wise thing and listen to your customers

**Comment by Klustre** [comment]
> > For the record, the title is wrong here, there is no security issue if Supabase just emailed the user to login if they attempt to register a second time.
> 
> I very much agree with this and I'll personally see what can be done on this point.

@hf I regularly get users that forgot that they signed up already and then wonder why they didn't receive a confirmation email after signing up again. The suggested solution above seems to be the best, considering the concerns. This issue is close to 3 years old and since you mentioned it comes up monthly I'd expect it to be a priority.

**Comment by appurist** [comment]
@Klustre What suggested solution are you referring to? (The .length "workaround" only works for some providers and isn't a general solution.) Do you mean for all developers to write a server-side function to use the Admin API to provide a new endpoint that does what registration should, and exposes the existence test anyway?

**Comment by Klustre** [comment]
> @Klustre What suggested solution are you referring to? (The .length "workaround" only works for some providers and isn't a general solution.) Do you mean for all developers to write a server-side function to use the Admin API to provide a new endpoint that does what registration should, and exposes the existence test anyway?

The solution suggested in the quote of my post.

**Comment by appurist** [comment]
@Klustre Thank you. I somehow read "solution" (Supabase-provided) as "workaround" (developer-coded). And yeah that was my quote so we are very much in agreement!

I think it's not a priority for SB because the developers disagree with it. But I don't see a downside to it.

**Comment by ddikman** [comment]
As mentioned above, it seems strange that the identities are returned in this case. If this decision is made for security it should return an empty list regardless.

When an existing account is used, the identities is empty but when a new user is created the identity gets set.

```
Seems like if we can just check if(data?.user?.identities?.length === 0){ then bad hackers could just read this thread, detect supabase, and check that too.
So, why not just return an error code instead?
```

I tried with a custom `createUser` function server side but reproducing the email confirmation flow used in the SDK seems difficult. As for the SQL stored procedure solution, would that not expose the table for anyone to simply iterate quite easily? The same is of course true with any endpoint created but I feel like an endpoint is easier to monitor and rate limit than a stored procedure but if anyone has a good idea here I'm eager to know.

**Comment by Rukannplatform** [comment]
For this issue i ended up doing a rpc function that checks if the user exist on auth.users table. 

```
CREATE OR REPLACE FUNCTION public.check_user_existence(user_email text)
 RETURNS TABLE(
  email_exist boolean
 )
 SECURITY DEFINER
 LANGUAGE sql AS
$func$

SELECT 
  EXISTS( 
  SELECT 1
  FROM auth.users
  WHERE email = user_email
  ) as email_exist
$func$;
```

**Comment by ricardosikic** [comment]
For this issue i ended up creating a rpc function that verifies if the email exist into auth.users table.

```
CREATE OR REPLACE FUNCTION public.check_user_existence(user_email text)
 RETURNS TABLE(
  email_exist boolean
 )
 SECURITY DEFINER
 LANGUAGE sql AS
$func$

SELECT 
  EXISTS( 
  SELECT 1
  FROM auth.users
  WHERE email = user_email
  ) as email_exist
$func$;
```

**Comment by oliverjhn** [comment]
> I'm not sure if this is the best approach, but this works well enough for me. I have an SQL function that checks if the user's email exists, which I call with RPC from my server action before the user is signed up.

That's the way! Are there any security or performance concerns?

**Comment by ddikman** [comment]
> > I'm not sure if this is the best approach, but this works well enough for me. I have an SQL function that checks if the user's email exists, which I call with RPC from my server action before the user is signed up.
> 
> That's the way! Are there any security or performance concerns?

As I mentioned above I wonder if this wouldn't make it very easy and quick for anyone with the anon key to simply iterate a bunch of emails to see if they exist or not. If you have an edge function you can add some rate limiting to it but perhaps there's a way to do that in the stored procedure as well?

**Comment by itsmartinhi** [comment]
Am I missing something? The current documentation clearly states that there should still be error returned in this case. Is the documentation out-of-date? https://supabase.com/docs/reference/javascript/auth-signup

**Comment by appurist** [comment]
@itsmartinhi It has been years now. Ant [said some things here](https://github.com/supabase/auth/issues/1517#issuecomment-2156222223) that I quoted but years later they are still 0 for 3 on those. I was 100% into Supabase until this issue, so it is one of the biggest disappointments in my 40-year professional career. Neither Firebase nor AppWrite have behavior like this, and are much more responsive to developers. Hint: I switched to AppWrite over this issue and it's very similar but much more predictable, usable, and feature-rich.

**Comment by kimhwanhoon** [comment]
I kinda figured out how to solve this. When you click the button from reset password mail, you'll be directed to a page with the code. 
1. with the code, you run `supabase.auth.exchangeCodeForSession`
2. then you will get `{data, error}` inside data, there will be redirectType which is set to `PASSWORD_RECOVERY`
3. redirect all unless they have specific redirect type
4. then only the ones who actually clicked to reset password from mail would be able to stay.

**Comment by nmcbride** [comment]
I'm new to supabase and am running into this same issue. It's very confusing.

If an account already exists, having a magic link automatically sent to the user would work fine but... it doesn't happen so...

**Comment by Klustre** [comment]
For anyone stumbling on this issue.

The docs for [`auth.signup`](https://supabase.com/docs/reference/javascript/auth-signup) mention:

> When **_either_** Confirm email **_or Confirm phone (even when phone provider is disabled) is disabled_**, the error message, `User already registered` is returned.

So disabling the `Enable phone confirmations` setting in the phone provider on the [auth providers page](https://supabase.com/dashboard/project/_/auth/providers) (while having `Confirm email` enabled in the email provider) should throw the `User already registered` error.

To have the same behaviour locally, add the following to `config.toml`:

```toml
[auth.sms]
enable_signup = false
enable_confirmations = false
```

Sources:
https://github.com/supabase/supabase/issues/18195#issuecomment-1903824780
https://github.com/supabase/supabase-js/issues/947

Thanks to @842u

**Comment by kedom1337** [comment]
This should probably get resolved ASAP. The biggest papercut I have had so far using Supabase and the only reason I would consider this only a papercut is due to the above "solution" by @Klustre / @842u

**Comment by Toukassay** [comment]
> For anyone stumbling on this issue.
> 
> The docs for [`auth.signup`](https://supabase.com/docs/reference/javascript/auth-signup) mention:
> 
> > When **_either_** Confirm email **_or Confirm phone (even when phone provider is disabled) is disabled_**, the error message, `User already registered` is returned.
> 
> So disabling the `Enable phone confirmations` setting in the phone provider on the [auth providers page](https://supabase.com/dashboard/project/_/auth/providers) (while having `Confirm email` enabled in the email provider) should throw the `User already registered` error.
> 
> To have the same behaviour locally, add the following to `config.toml`:
> 
> ```toml
> [auth.sms]
> enable_signup = false
> enable_confirmations = false
> ```
> 
> Sources: [supabase/supabase#18195 (comment)](https://github.com/supabase/supabase/issues/18195#issuecomment-1903824780) [supabase/supabase-js#947](https://github.com/supabase/supabase-js/issues/947)
> 
> Thanks to @842u

Funny but it works :D

**Comment by spraddles** [comment]
It's Jan 2025 and now 3+ years since this issue was raised. 
Seriously guys @ `supabase` answer this simple riddle for me:
- if a user wants to use a social login to my app
- how can I onboard them
- if I can't tell that they're an existing customer
- so I am forced to make them go through my extra onboarding screens again?

Unbelievable.

**Comment by nmerchant** [comment]
There's a simple way to implement this check yourself.

Create a database function called `get_user_id_by_email`:

```
BEGIN
  RETURN QUERY SELECT au.id FROM auth.users au WHERE au.email = $1;
END;
```

and call it in your app like this:

```
const { data: emailExists, error } = await supabase.rpc("get_user_id_by_email", { email: schema.email });
    if (emailExists?.length) {
        console.log('Email already exists');
    }
```

**Comment by spraddles** [comment]
> There's a simple way to implement this check yourself.
> 
> Create a database function called `get_user_id_by_email`:
> 
> ```
> BEGIN
>   RETURN QUERY SELECT au.id FROM auth.users au WHERE au.email = $1;
> END;
> ```
> 
> and call it in your app like this:
> 
> ```
> const { data: emailExists, error } = await supabase.rpc("get_user_id_by_email", { email: "foo@example.com" });
> if (emailExists?.length) console.log('Email already exists');
> ```

I’ve already done this but you can’t stop a User from being created via Social auth (like Google, Apple etc) as you can’t check an email from Social prior to the user logging in, only afterwards. And by that time the User is already created, which doesn’t make any sense.

**Comment by syberkitten** [comment]
I just came up to this error by myself, and was astounded to find this as an Open issue for 4 years.
How come in my local development it does work?

I do get back:
{"code":"user_already_exists","message":"User already registered"} 


Can someone explain how come Dev version works properly and Prod now?

**Comment by silentworks** [comment]
@syberkitten It's stated in one of the replies above, it has to do with your settings. If you have confirm email turned off you will get the error message you have above, if it is turned on then you get back the dummy data.

**Comment by nikitph** [comment]
Another victim here. And just to be clear this has nothing to do with email confirmation but the enable phone confirmation under phone provider even if we are not using it. our workaround is to just disable that and then the user_already exists starts showing up. Very peculiar .

**Comment by MiguelCedric** [comment]
> [@syberkitten](https://github.com/syberkitten) It's stated in one of the replies above, it has to do with your settings. If you have confirm email turned off you will get the error message you have above, if it is turned on then you get back the dummy data.

Hi, you can turn off confirm phone instead and leave email confirmed it will work.

**Comment by BilalTariq01** [comment]
Disabling the phone confirmation (whether you are using it or not) will work.

**Comment by ShahmeerBaig1** [comment]
> Disabling the phone confirmation (whether you are using it or not) will work.
 
Yeah this worked!

**Comment by Fingertips18** [comment]
It's weird that it works only after disabling Enable phone confirmations in the phone provider, even though it was already disabled. I also tried signing up with unique details but using the same email—expected it to trigger the email_exists error code, but instead it returns user_already_exists. So strange...

**Comment by holahoon** [comment]
This was driving me nuts. I had to disable Confirm email and then re-enable it just to trigger the error, even though Enable phone confirmations is turned off. Something’s definitely not working right 🤔
I'm testing this in dev mode for now, but I’ll need to test it in the Pro plan’s production version too. It feels like I’m flying blind.

**Comment by ogirrk** [comment]
In general, confirm email is desired for production. Now that the feature [Auth Hooks](https://supabase.com/docs/guides/auth/auth-hooks) has been introduced, wouldn't the following be a simple solution (probably does not cover every use case)?

Currently, if Supabase Auth (gotrue) detects the email is already registered, the "Before User Created Hook" is not executed. Why not just let the "Before User Created Hook" (or introduce a specific hook, e.g., "Signup Hook") to be executed before Supabase Auth checks for any existing registered email?
Within the secured Postgres function evoked by the hook, one can check for already registered email or anything they want to check (and return appropriate error messages to the front-end), or leave it and return '{}' to let Supabase Auth deals with it.

**Comment by appurist** [comment]
I needed a Postgres provider with authentication, so I checked this issue, but it _still_ hasn't been resolved. So again, my project will be hosted elsewhere. (I've decided to go with n-e-o-n for anyone else in the same boat who needs real Postgres.)
